spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java

```diff
@@ -272,11 +275,31 @@
 	/**
 	 * Finds objects by using a Cloud Datastore query. If the query is a key-query, then keys are
 	 * returned.
+	 * Resulting Slice can be used to get a Pageable for the next page or to determine if next page exists.
 	 * @param query the query to execute.
 	 * @param entityClass the type of object to retrieve.
+	 * @param pageable that indicates page number and page size
 	 * @param <T> the type of object to retrieve.
-	 * @return a list of the objects found. If no keys could be found the list will be
-	 * empty.
+	 * @return a Slice containing found objects
+	 */
+	<T> Slice<?> queryKeysOrEntitiesSlice(StructuredQuery query, Class<T> entityClass, Pageable pageable);
```

Comment:
Can you remind me, what is the benefit of combining `queryEntities()` and `queryKeys()` into one method?
Sonar is identifying this as a "code smell".

Reply:
The code would be identical for these cases, so we are avoiding duplication this way.

Reply:
Let me see if I can refactor it to avoid duplication.